### ISSUE TITLE
History browser: Browse and restore old file versions 

### ISSUE THREAD

**Issue by binwiederhier** [syncany/syncany#333]
One of Syncany's strengths is that we keep the history of the files we manage, so we can easily go back to previous versions, but we can also browse a managed folder at a previous point in time which is something that (to the best of my knowledge) non of the others can do. 

The PoC web interface [shows that "time browser" with the time slider](http://youtu.be/2NPVffywZVs?t=43s):

![2014123013 58 16_selection_001](https://cloud.githubusercontent.com/assets/664597/5578664/f51d2e5e-902b-11e4-9943-af5e786206f6.png)

![2014123014 03 42_one_001](https://cloud.githubusercontent.com/assets/664597/5578707/caa8ecfc-902c-11e4-8439-a5480a338e9f.png)

In my opinion, it would be nice to implement such a file browser with time slider and restore functionality at some point in the desktop GUI. I have a vague idea of how this should look like, but I think some discussion is useful before we implement anything. I would basically just implement what you see in the web interface for the GUI: A separate window with a time slider and a "previous versions" button.

SparkleShare also has a "previous versions" window which is linked to the "Recent changes" window (see #332):

![2014123014 00 13_selection_001](https://cloud.githubusercontent.com/assets/664597/5578714/fbbd6700-902c-11e4-9eee-31e6d72edd20.png)

Opinions?

**Comment by binwiederhier** [comment]
Sneak peek of the history browser: 

Video:
https://www.youtube.com/watch?v=JoISUKP90wI&feature=youtu.be

Screenshot:
![2015010716 31 37_selection_001](https://cloud.githubusercontent.com/assets/664597/5647953/af2bd390-968a-11e4-9e8e-4e1b209ba30b.png)

**Comment by pimotte** [comment]
Looking good! I think it would be a big plus if the folded/unfolded status of folders remains (as far as possible) when sliding through history. (I can imagine if that's too much, though).

**Comment by cr0** [comment]
I would rather see that as a requirement.

**Comment by binwiederhier** [comment]
How about that: 
https://www.youtube.com/watch?v=wbrFIEQ_5Xo&feature=youtu.be

**Comment by binwiederhier** [comment]
I had to alter `sy ls` (aka the `LsOperation`) a bit to make this work, but it can now also select by file history ID, so all the better. I have noticed though that `sy ls` is slow with large databases, because the SQL statement is rather inefficient (unindexed queries, `like`s and such). I'll make another issue for that.

**Comment by binwiederhier** [comment]
Made it a bit prettier!

![2015010801 53 23_selection_001](https://cloud.githubusercontent.com/assets/664597/5656219/4cfb3428-96d9-11e4-9904-a46e5e6a2c2d.png)
![2015010801 53 42_selection_001](https://cloud.githubusercontent.com/assets/664597/5656220/5024abf2-96d9-11e4-91c8-e0e78cb7b513.png)

**Comment by cr0** [comment]
looks good.

Perhaps we should add the last modified date in the overview as well. Use case: I want to restore the previous version of a file but I've no idea when I changed it (a day ago or two?)

**Comment by binwiederhier** [comment]
How about that?

![2015010813 15 26_selection_001](https://cloud.githubusercontent.com/assets/664597/5662364/7af526ae-9738-11e4-9940-2714814a612a.png)

**Comment by binwiederhier** [comment]
I'll make another view like this so that we can see what has changed.

![fbbd6700-902c-11e4-9eee-31e6d72edd20](https://cloud.githubusercontent.com/assets/664597/5662474/bc55896c-9739-11e4-8439-32d9f2e5a27d.png)

**Comment by binwiederhier** [comment]
Not entirely finished. But looking goood.

![2015010816 50 18_selection_001](https://cloud.githubusercontent.com/assets/664597/5665139/987ce5b8-9756-11e4-910f-6a6482088e97.png)

![2015010816 50 33_selection_001](https://cloud.githubusercontent.com/assets/664597/5665140/9f3ce448-9756-11e4-94b8-ebe68e45120c.png)

**Comment by binwiederhier** [comment]
Updated:
https://www.youtube.com/watch?v=U8ponntr_HA&feature=youtu.be

**Comment by cr0** [comment]
Are you planning to show which machine did the actual change operation? Imho the history view would greatly benefit from knowing who changed a file.

**Comment by binwiederhier** [comment]
Unfortunately we currently only have the machine name, not the display name. We could/should add the displayName to the databaseversion header to be able to display it.  I don't think there is any value in displaying the machineName.

**Comment by binwiederhier** [comment]
Implemented re-loading of tab panels (do not load all at once). Also, only 10 file entries are displayed at a time. If more might be available, a 'More ...' label is displayed. 

Still to do:

- [ ] File entries should be clickable and lead to the detailed file view (or the file tree?!)
- [ ] The 'More ...' entry should be clickable and reload the entire database entry.

![2015010919 35 48_selection_001](https://cloud.githubusercontent.com/assets/664597/5685075/2898fdae-9837-11e4-9f2d-e3962e722c9e.png)

**Comment by binwiederhier** [comment]
Completely refactored. Works like a charm now. I also added a nice loading panel. Still many missing things tough ...

- [ ] When not all tabs are loaded and the date slider is moves "too far" left (= not-yet-loaded date/tab), no tab is highlighted; and tabs are not loaded; we should load them.
- [ ] Restore works, but has not progress indicator
- [ ] Refactor LogRequest/Response to return `<fileVersion>`s instead of `ChangeSet`
- [ ] (optional) If GUI/daemon are in same JVM, don't use websockets; instead, use eventbus only

**Comment by binwiederhier** [comment]
**Context menu**
![2015012523 20 13_selection_001](https://cloud.githubusercontent.com/assets/664597/5893771/c213da0e-a4e8-11e4-8819-185e47a36c4d.png)

**More link expands the view**
![2015012523 19 05_selection_001](https://cloud.githubusercontent.com/assets/664597/5893772/c4ff8f56-a4e8-11e4-8302-4eee61eb4ae1.png)
![2015012523 19 18_selection_001](https://cloud.githubusercontent.com/assets/664597/5893773/c698259e-a4e8-11e4-8f31-f6b07d6282d7.png)

**Loading screen:**
![2015012523 17 45_selection_001](https://cloud.githubusercontent.com/assets/664597/5893774/c9b27090-a4e8-11e4-918a-5ca835be630e.png)

**Comment by gcala** [comment]
Are you planning to add direct access to files, I mean open in external application a file listed in the history browser?

**Comment by binwiederhier** [comment]
@gcala Interestingly, we *just* had this discussion on IRC, so here's an excerpt:

```
<A> hi, is it possible to open a given version of a file?
    (f.i., a text file's version in text editor). I presume it needs some
    virtual file system... Thanks
<B> A: as far as i know, no. There's the "Restore" operation,
    which can restore a file to a previous version, so if you backup
    the current version you can use that
<C> use sy ls --versions --group to list the versions, and then sy restore
    --revision=X <file-id> to restore version X
<A> B: ok; I think that could be useful to preview the state of the file 
    and decide which version to restore
<C> A: previewing would mean we'd have to restore the file to a temp
    folder. we don't have all the actual data locally, just the metadata
<C> if it's just a text file that wouldnt be a problem. something like
    "sy preview ..", but for large files it'd be difficult
```

Even though I (C) said it's difficult, I actually remembered now that I implemented that and the code should still be there. It still means having to retrieve/decrypt/assemble the entire file, but it can definitely work. We should add that as a separate issue though. @gcala Would you mind opening a new issue for that and link to this comment? Thanks!

**Comment by gcala** [comment]
@binwiederhier no problem at all

**Comment by binwiederhier** [comment]
I think it might be done. Restore progress works. It might need some polishing here and there and a lot of testing. The two missing issues from above will be solved in two other issues.

![2015020601 24 18_selection_001](https://cloud.githubusercontent.com/assets/664597/6073025/bb086840-ada0-11e4-90b5-b829bc110f77.png)
![2015020601 24 42_selection_001](https://cloud.githubusercontent.com/assets/664597/6073031/c73b6018-ada0-11e4-9a95-8ecabbe54316.png)

**Comment by binwiederhier** [comment]
Merged. 
*Can I get a whooo-whoo?!*

**Comment by cr0** [comment]
Bad news:
Syncany, or more precisely the JVM is crashing once I close the history browser:

```
13-2-15 11:43:12.691 | HistoryDialog   | main           | INFO : History dialog: Disposing dialog ...
13-2-15 11:43:12.693 | InternalEventBu | main           | INFO : Event bus 'GuiEventBus': Unregistering HistoryDialog (org.syncany.gui.history.HistoryDialog@258e2e41) ...
13-2-15 11:43:12.695 | InternalEventBu | main           | INFO : Event bus 'GuiEventBus': Unregistering MainPanel (MainPanel {}) ...
13-2-15 11:43:12.695 | InternalEventBu | main           | INFO : Event bus 'GuiEventBus': Unregistering LogComposite (LogComposite {}) ...
13-2-15 11:43:12.714 | FileTreeComposi | main           | INFO : Tree: Disposing tree ...
13-2-15 11:43:12.715 | InternalEventBu | main           | INFO : Event bus 'GuiEventBus': Unregistering FileTreeComposite (FileTreeComposite {}) ...
13-2-15 11:43:12.716 | DetailPanel     | main           | INFO : Detail panel: Disposing panel ...
13-2-15 11:43:12.716 | InternalEventBu | main           | INFO : Event bus 'GuiEventBus': Unregistering DetailPanel (DetailPanel {}) ...
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fff96dba8ee, pid=35500, tid=1823
#
# JRE version: Java(TM) SE Runtime Environment (8.0_40-b02) (build 1.8.0_40-ea-b02)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.40-b05 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [vImage+0x268ee]  vPremultipliedAlphaBlendWithPermute_ARGB8888+0x66e
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/chr/Work/syncany-plugin-gui/hs_err_pid35500.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.sun.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
[1]    35500 abort      sy gui -d
``

**Comment by binwiederhier** [comment]
This is most likely an SWT dispose() issue. The application always crashes when you're accessing a disposed SWT widget. Normally I do disposed-checks everywhere, so this shouldn't happen. 

Unfortunately, since this does not happen on my machine, I cannot debug it. Can you try and track it down. If you need help or have questions regarding SWT, let me know. I'm available.

**Comment by binwiederhier** [comment]
Is this reproducible? Does it happen the exact same way every time? 

From the log, I'd say that `DetailPanel.dispose()` terminates just fine, but `windowShell.dispose()` in `HistoryDialog.dispose()` causes problems. Can you add a couple of printlns to track down the exact line that causes problems? Can you play around with switching `Display.getDefault().syncExec()` to `Display.getDefault().asyncExec()` in the dispose methods?

I know that you have the subfoldering-issue to tackle, but this is even higher prio, because **it blocks the next release**. Thanks!

**Comment by cr0** [comment]
It happens all the time, thus yes, it is reproducible. I'll try to find some spare time in the evening.

**Comment by cr0** [comment]
Found the reason for the crash with the help of a dump:

```
Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  org.eclipse.swt.internal.cocoa.OS.objc_msgSend(JJLorg/eclipse/swt/internal/cocoa/NSRect;Lorg/eclipse/swt/internal/cocoa/NSRect;JD)J+0
j  org.eclipse.swt.internal.cocoa.NSImage.drawInRect(Lorg/eclipse/swt/internal/cocoa/NSRect;Lorg/eclipse/swt/internal/cocoa/NSRect;JD)V+12
j  org.eclipse.swt.graphics.GC.drawImage(Lorg/eclipse/swt/graphics/Image;IIIIIIIIZ)V+284
j  org.eclipse.swt.graphics.GC.drawImage(Lorg/eclipse/swt/graphics/Image;II)V+42
j  org.syncany.gui.history.ImageComposite$AnimatedGifComposite$3.run()V+101
J 5236 C1 org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Z)Z (133 bytes) @ 0x000000010662ade4 [0x000000010662aa80+0x364]
J 5203 C1 org.eclipse.swt.widgets.Display.readAndDispatch()Z (247 bytes) @ 0x0000000106f8ba7c [0x0000000106f8b600+0x47c]
j  org.syncany.operations.gui.GuiOperation.startEventDispatchLoop()V+15
j  org.syncany.operations.gui.GuiOperation.execute()Lorg/syncany/operations/OperationResult;+44
j  org.syncany.cli.GuiCommand.execute([Ljava/lang/String;)I+16
j  org.syncany.cli.CommandLineClient.runLocally(Lorg/syncany/cli/Command;[Ljava/lang/String;)I+15
j  org.syncany.cli.CommandLineClient.runCommand(Lorg/syncany/cli/Command;Ljava/lang/String;[Ljava/lang/String;)I+168
j  org.syncany.cli.CommandLineClient.start()I+385
j  org.syncany.Syncany.main([Ljava/lang/String;)V+8
v  ~StubRoutines::call_stub
```

It's actually your animation's fault because as far as I comprehend `ImageComposite$AnimatedGifComposite.nextImage` still tries to draw in an already disposed area. More precisely [`canvas.redraw()`](https://github.com/syncany/syncany-plugin-gui/blob/develop/src/main/java/org/syncany/gui/history/ImageComposite.java#L179) is causing the crash. However, the bar seems does not seem to be animated on OS X anyway.

**Comment by cr0** [comment]
Why did you choose to use a GIF for this kind of loading animation and not [SWT's progress bar with indeterminate style](http://www.java2s.com/Tutorial/Java/0280__SWT/UsingtheSWTINDETERMINATEstyletheoperationishappeningbutdoesntknowwhenitllfinish.htm)?

**Comment by binwiederhier** [comment]
We could use the progress bar in this context, but we're also using the `ImageComposite` in the restore panel. When you click the restore button, there is a rotating loading icon.  In that sense, we need the image composite.

**Comment by cr0** [comment]
I added a few comments to the code:

```		
private void nextImage() {
	Display.getDefault().syncExec(new Runnable() {
		public void run() {
			try {
				logger.log(Level.FINE, "nextImage, waiting");
				synchronized(ImageComposite.this) {
					logger.log(Level.FINE, "nextImage, go");
					imageNumber = (imageNumber == loader.data.length - 1) ? 0 : imageNumber + 1;

					ImageData nextFrameData = loader.data[imageNumber];
					Image frameImage = new Image(Display.getDefault(), nextFrameData);

					gc.drawImage(frameImage, nextFrameData.x, nextFrameData.y);

					frameImage.dispose();
					logger.log(Level.FINE, "nextImage, dispose");
					canvas.redraw();
					logger.log(Level.FINE, "nextImage, redraw");
				}
			}
			catch (Exception e) {
				// Might be disposed. We don't care, because another image
				// is in the process of being initialized.
			}
		}
	});
}

@Override
public void dispose() {
	logger.log(Level.FINE, "Dispose, waiting");
	synchronized(this) {
		logger.log(Level.FINE, "Dispose, go");
		imageTimer.cancel();
		logger.log(Level.FINE, "Dispose, timer canceled");
		super.dispose();
		logger.log(Level.FINE, "Dispose, disposing");
	}
}
```

Resulting log:

```
22-2-15 13:35:37.453 | org.syncany.gui | main           | FINE : nextImage, go
22-2-15 13:35:37.453 | org.syncany.gui | main           | FINE : nextImage, dispose
22-2-15 13:35:37.542 | org.syncany.gui | main           | FINE : nextImage, waiting
22-2-15 13:35:37.542 | org.syncany.gui | main           | FINE : nextImage, go
22-2-15 13:35:37.543 | org.syncany.gui | main           | FINE : nextImage, dispose
22-2-15 13:35:37.575 | HistoryDialog   | main           | INFO : History dialog: Disposing dialog ...
22-2-15 13:35:37.575 | InternalEventBu | main           | INFO : Event bus 'GuiEventBus': Unregistering HistoryDialog (org.syncany.gui.history.HistoryDialog@19d37183) ...
mainPanel.dispose()
22-2-15 13:35:37.579 | InternalEventBu | main           | INFO : Event bus 'GuiEventBus': Unregistering MainPanel (MainPanel {}) ...
22-2-15 13:35:37.580 | InternalEventBu | main           | INFO : Event bus 'GuiEventBus': Unregistering LogComposite (LogComposite {}) ...
22-2-15 13:35:37.593 | FileTreeComposi | main           | INFO : Tree: Disposing tree ...
22-2-15 13:35:37.594 | InternalEventBu | main           | INFO : Event bus 'GuiEventBus': Unregistering FileTreeComposite (FileTreeComposite {}) ...
detailPanel.dispose()
22-2-15 13:35:37.595 | DetailPanel     | main           | INFO : Detail panel: Disposing panel ...
22-2-15 13:35:37.596 | InternalEventBu | main           | INFO : Event bus 'GuiEventBus': Unregistering DetailPanel (DetailPanel {}) ...
windowShell.dispose()
22-2-15 13:35:37.641 | org.syncany.gui | main           | FINE : nextImage, waiting
22-2-15 13:35:37.641 | org.syncany.gui | main           | FINE : nextImage, go
<CRASH>
```

I've almost no experience with GUI development (with SWT) so I'm afraid that I'm not much of a help here.

**Comment by binwiederhier** [comment]
I've added a check for the disposed status of canvas. 

(a) Can you check if that makes a difference?
(b) If that doesn't work, can you maybe remove the `frameImage.dispose();`?
(c) If that doesn't work, can you check if changing the `nextImage()` call from `Display.getDefault().syncExec()` to ``Display.getDefault().asyncExec()` changes things?
(d) If that doesn't work, can you check if [this example](http://stackoverflow.com/questions/13479833/java-swt-animated-gif) works? 
(e) If that doesn't work, it might be a cocoa bug. Then we should implement a OSX workaround by doing something like: 

```java
public void setAnimatedImage(...) {
  if (Env.isOSX()) {
     setImage(...);
   }
   else {
     // Good stuff
   }
}
```

**Comment by cr0** [comment]
I'll provide feedback tomorrow, probably in the evening.

--

On 22. Februar 2015 22:46:28 MEZ, Philipp C. Heckel <notifications@github.com> wrote:I've added a check for the disposed status of canvas.   (a) Can you check if that makes a difference? (b) If that doesn't work, can you maybe remove the frameImage.dispose(); ? (c) If that doesn't work, can you check if changing the nextImage()  call from Display.getDefault().syncExec()  to `Display.getDefault().asyncExec()  changes things? (d) If that doesn't work, can you check if this example works?  (e) If that doesn't work, it might be a cocoa bug. Then we should implement a OSX workaround by doing something like:   public void setAnimatedImage(...) { if (Env.isOSX()) { setImage(...); } else { // Good stuff } }   —Reply to this email directly or view it on GitHub.

**Comment by binwiederhier** [comment]
:+1:

**Comment by binwiederhier** [comment]
Not get all annoying, but this item is the sole thing preventing us from releasing 0.4.3. :)

**Comment by binwiederhier** [comment]
For reference: 0.4.3-alpha went out with the history browser, but only for non-OSX systems. I removed the workaround again in the latest develop though.

**Comment by cr0** [comment]
So with a week delay (sorry about that):

a) not working
b) not working
c) not working
d) http://stackoverflow.com/a/28492105/1288692 works. If you want me to try the Syncany code, could you provide a minimal example please?

I think removing the history browser is a bit extreme we should simply skip the `canvas.redraw` call. In that case the browser works fine.

**Comment by cr0** [comment]
[1c6639e](https://github.com/syncany/syncany-plugin-gui/commit/1c6639e86bf9607967d0393002b52b38f0fe53c8?w=1#diff-124e38bd48a141886976700ef11b3507R103) fixes the crash.

**Comment by binwiederhier** [comment]
Yeyy. I'll release a new version tonight (if i remember it :))
Good work!
